src/fileformats/ocd_georef.h

```diff
@@ -0,0 +1,115 @@
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    OpenOrienteering is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with OpenOrienteering.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef OCD_GEOREF_H
+#define OCD_GEOREF_H
+
+#include <functional>
+#include <utility>
+#include <cmath>
+#include <limits>
+
+#include <QString>
+
+#include "core/georeferencing.h"
+
+namespace OpenOrienteering {
+
+class OcdGeoref
+{
+	Q_DECLARE_TR_FUNCTIONS(OpenOrienteering::OcdGeoref)
+
+public:
+	struct OcdStringFields
+	{
+		// Named after OCD string fields. Structure element name
+		// corresponds to the code and its value is the field value.
+		double a;
+		int m, x, y, i;
+
+		OcdStringFields() :
+		    a(std::nan("OCD")),
+		    m(-1),
+		    x(std::numeric_limits<int>::min()),
+		    y(std::numeric_limits<int>::min()),
+		    i(-1)
+		{ }
+	};
+
+	/**
+	 * @brief Fill in provided georef with data from type 1039 string.
+	 * @param georef Reference to a Georeferencing object.
+	 * @param param_struct Data extracted from OCD type 1039 string.
+	 * @param warning_handler Function to handle import warnings.
+	 */
+	static void georefFromFields(Georeferencing& georef,
+	                             const OcdStringFields& param_struct,
+	                             const std::function<void (const QString&)>& warning_handler);
+
+	/**
+	 * @brief fieldsFromGeoref
+	 * @param georef Source Georeferencing.
+	 * @param warning_handler Function to handle conversion warnings.
+	 * @return Data for OCD type 1039 string compilation.
+	 */
+	static OcdStringFields fieldsFromGeoref(const Georeferencing& georef,
+	                                           const std::function<void (const QString&)>& warning_handler);
+
+protected:
+	enum struct MapperCrs;
+	static const std::map<const MapperCrs, const QString> crs_string_map;
+	static const std::map<const QString, const MapperCrs> crs_id_map;
+
+	struct MapperCrsRecord
+	{
+		const MapperCrs crs_id; // CRS id, same strings as CRSTemplateRegistry::find() consumes
+		const int crs_param; // the parameter to be passed through CRSTemplateRegistry::specValues()
+	};
+
+	struct OcdRecord
+	{
+		const int ocd_grid_id; // OCD grid/zone numbers
+		const int ocd_zone_id;
+	};
+
+	/// Imports string 1039 field i.
+	static void applyGridAndZone(Georeferencing& georef,
+	                             const int combined_ocd_grid_zone,
+	                             const std::function<void(const QString&)>& warning_handler);
+
+	/**
+	 * @brief Plain translation from OCD grid/zone combination
+	 *        into possible Mapper schemes.
```

Comment:
BTW we don't need `@brief` because doxygen runs with autobrief activated.

Reply:
Fixed.